// @oracle/oraclejet-templates/navdrawer-ts/hybrid/src/ts/appController.ts
/*
 * Your application specific code will go here
 */
import * as ko from "knockout";
import CoreRouter = require("ojs/ojcorerouter");
import ModuleRouterAdapter = require("ojs/ojmodulerouter-adapter");
import KnockoutRouterAdapter = require("ojs/ojknockoutrouteradapter");
import UrlParamAdapter = require("ojs/ojurlparamadapter");
import * as ModuleElementUtils from "ojs/ojmodule-element-utils";
import ArrayDataProvider = require("ojs/ojarraydataprovider");
import * as KnockoutTemplateUtils from "ojs/ojknockouttemplateutils";
import "ojs/ojknockout";
import "ojs/ojmodule-element";
import * as OffcanvasUtils from "ojs/ojoffcanvas";
import { ojNavigationList } from "ojs/ojnavigationlist";
import { ModuleViewModel } from "ojs/ojmodule-element";
import Context = require("ojs/ojcontext");

interface CoreRouterDetail {
  label: string;
  iconClass: string;
};

class HeaderModel implements ModuleViewModel {
  pageTitle: string;
  toggleDrawer: Function;
  adjustContentPadding: Function;

  constructor({ pageTitle, toggleDrawer, adjustContentPadding } : {
    pageTitle: string;
    toggleDrawer: Function;
    adjustContentPadding: Function;
  }) {
    this.pageTitle = pageTitle;
    this.toggleDrawer = toggleDrawer;
    this.adjustContentPadding = adjustContentPadding;
  }

  transitionCompleted(): null {
    // adjust content padding after header bindings have been applied
    this.adjustContentPadding();
    return null;
  }
}

class RootViewModel {
  KnockoutTemplateUtils: typeof KnockoutTemplateUtils;
  manner: ko.Observable<string>;
  message: ko.Observable<string|undefined>;
  waitForAnnouncement: boolean;
  navDrawerOn: boolean;
  router: CoreRouter<CoreRouterDetail>|undefined;
  moduleAdapter: ModuleRouterAdapter<CoreRouterDetail>;
  navDataProvider: ojNavigationList<string, CoreRouter.CoreRouterState<CoreRouterDetail>>["data"];
  selection: KnockoutRouterAdapter<CoreRouterDetail>;

  constructor() {
      this.KnockoutTemplateUtils = KnockoutTemplateUtils;

      // handle announcements sent when pages change, for Accessibility.
      this.manner = ko.observable("polite");
      this.message = ko.observable();
      this.waitForAnnouncement = false;
      this.navDrawerOn = false;

      let globalBodyElement: HTMLElement = document.getElementById("globalBody") as HTMLElement;
      globalBodyElement.addEventListener("announce", this.announcementHandler, false);

      const navData = [
        { path: "", redirect: "dashboard" },
        { path: "dashboard", detail: { label: "Dashboard", iconClass: "oj-ux-ico-bar-chart" } },
        { path: "incidents", detail: { label: "Incidents", iconClass: "oj-ux-ico-fire" } },
        { path: "customers", detail: { label: "Customers", iconClass: "oj-ux-ico-contact-group" } },
        { path: "profile", detail: { label: "Profile", iconClass: "oj-ux-ico-contact" } },
        { path: "about", detail: { label: "About", iconClass: "oj-ux-ico-information-s" } }
      ];
      // router setup
      const router = new CoreRouter(navData, {
        urlAdapter: new UrlParamAdapter()
      });
      router.sync();

      this.moduleAdapter = new ModuleRouterAdapter(router);

      this.selection = new KnockoutRouterAdapter(router);

      // Setup the navDataProvider with the routes, excluding the first redirected
      // route.
      this.navDataProvider = new ArrayDataProvider(navData.slice(1), {keyAttributes: "path"});

      // release the application bootstrap busy state
      Context.getPageContext().getBusyContext().applicationBootstrapComplete();        
    }

    /*
      @waitForAnnouncement - set to true when the announcement is happening.
      If the nav-drawer is ON, it is reset to false in "ojclose" event handler of nav-drawer.
      If the nav-drawer is OFF, then the flag is reset here itthis in the timeout callback.
    */
    announcementHandler = (event: any): void => {
      this.waitForAnnouncement = true;
      this.message(event.detail.message);
      this.manner(event.detail.manner);
      if (!this.navDrawerOn) {
        this.waitForAnnouncement = false;
      }
    }

    toggleDrawer = (): Promise<boolean> => {
      this.navDrawerOn = true;
      return OffcanvasUtils.toggle({selector: "#navDrawer", modality: "modal", content: "#pageContent"});
    }

    getHeaderModel(): HeaderModel {
      let pageTitle: string = this.selection.state().detail.label;
      let toggleDrawer: Function = this.toggleDrawer;
      let adjustContentPadding: Function = this.adjustContentPadding;
      return new HeaderModel( { pageTitle, toggleDrawer, adjustContentPadding } );
    }

    // method for adjusting the content area top/bottom paddings to avoid overlap with any fixed regions.
    // this method should be called whenever your fixed region height may change.  The application
    // can also adjust content paddings with css classes if the fixed region height is not changing between
    // views.
    adjustContentPadding(): void {
      let topElem: HTMLElement = document.getElementsByClassName("oj-applayout-fixed-top")[0] as HTMLElement;
      let contentElem: HTMLElement = document.getElementsByClassName("oj-applayout-content")[0] as HTMLElement;
      let bottomElem: HTMLElement = document.getElementsByClassName("oj-applayout-fixed-bottom")[0] as HTMLElement;

      if (topElem) {
        contentElem.style.paddingTop = `${topElem.offsetHeight}px`;
      }
      if (bottomElem) {
        contentElem.style.paddingBottom = `${bottomElem.offsetHeight}px`;
      }
      // add oj-complete marker class to signal that the content area can be unhidden.
      // see the override.css file to see when the content area is hidden.
      contentElem.classList.add("oj-complete");
    }
}

export default new RootViewModel();
