// @oracle/oraclejet-preact/es/UNSAFE_Badge.js
/* @oracle/oraclejet-preact: 14.0.4 */
import { jsx } from 'preact/jsx-runtime';
import { classNames } from './utils/UNSAFE_classNames.js';
import "./UNSAFE_Badge.css";

const styles = {
  base: "_17ikhni",
  //right-aligned with the right edge of another component for ltr direction
  //left-aligned for rtl direction pending by
  //https://jira.oraclecorp.com/jira/browse/JET-47572: need an api for rtl
  end: "lrg2a9",
  neutral: "hyi7n1",
  danger: "_7oezb1",
  warning: "_1w87aj9",
  success: "_1v907y6",
  info: "lo0ex3",
  neutralSubtle: "_1ye0x4k",
  dangerSubtle: "_195p35l",
  warningSubtle: "uofy8f",
  successSubtle: "oxtisk",
  infoSubtle: "glnt6r",
  childrenMargins: "_11t87zl"
};
function Badge({
  variant = 'neutral',
  size = 'md',
  edge = 'none',
  children
}) {
  const textClasses = classNames([styles.base, size === 'sm' && 'oj-badge-sm', edge === 'end' && styles.end, styles[variant], styles.childrenMargins]);
  return jsx("span", Object.assign({
    class: textClasses
  }, {
    children: children
  }));
}

export { Badge };
/*  */
