// @oracle/oraclejet-preact/es/UNSAFE_Modal.js
/* @oracle/oraclejet-preact: 14.0.4 */
import { jsx, jsxs } from 'preact/jsx-runtime';
import { useRef, useCallback, useEffect, forwardRef } from 'preact/compat';
import { FocusTrap } from './UNSAFE_FocusTrap.js';
import { Layer } from './UNSAFE_Layer.js';
import "./UNSAFE_Modal.css";
import 'preact';
import './utils/UNSAFE_classNames.js';
import './utils/PRIVATE_tabbableUtils.js';
import './_curry1-f2673bab.js';
import './_curry2-d7587d88.js';
import './_curry3-6b8e8079.js';

/**
 *
 * @param props
 */
const useModal = (props) => {
    const { isOpen, onBackdropClick: onBackdropClickProp, } = props;
    const modalRef = useRef(null);
    const mouseDownTarget = useRef(null);
    const onMouseDown = useCallback((event) => {
        if (modalRef.current === event.target) {
            mouseDownTarget.current = event.target;
        }
    }, []);
    const onBackdropClick = useCallback((event) => {
        if (modalRef.current === event.target) {
            event.stopPropagation();
            // event should originate on the same target
            if (mouseDownTarget.current === event.target) {
                onBackdropClickProp === null || onBackdropClickProp === void 0 ? void 0 : onBackdropClickProp();
            }
        }
    }, [onBackdropClickProp]);
    useEffect(() => {
        const modal = modalRef.current;
        if (modal) {
            modal.addEventListener('mousedown', onMouseDown);
            modal.addEventListener('click', onBackdropClick);
            return () => {
                modal.removeEventListener('mousedown', onMouseDown);
                modal.removeEventListener('click', onBackdropClick);
            };
        }
        return () => { };
    }, [isOpen, onMouseDown, onBackdropClick]);
    return {
        modalRef,
    };
};

const styles$2 = {
  base: "_12yz8ja"
};

const ModalBackdrop = () => jsx("div", {
  class: styles$2.base
});

const styles$1 = {
  base: "_1vxwng5"
};
const ModalContainer = forwardRef(({
  children
}, ref) => jsx("div", Object.assign({
  class: styles$1.base,
  ref: ref
}, {
  children: children
})));
ModalContainer.displayName = 'Forwarded<ModalContainer>';

const styles = {
  base: "_1wzau58"
};

const ModalWrapper = ({
  children
}) => jsx("div", Object.assign({
  class: styles.base
}, {
  children: children
}));

const Modal = ({
  autoFocusRef,
  children,
  isOpen,
  onBackdropClick,
  restoreFocusRef
}) => {
  const {
    modalRef
  } = useModal({
    isOpen,
    onBackdropClick
  });
  return isOpen ? jsx(Layer, {
    children: jsxs(ModalWrapper, {
      children: [jsx(ModalBackdrop, {}), jsx(ModalContainer, Object.assign({
        ref: modalRef
      }, {
        children: jsx(FocusTrap, Object.assign({
          autoFocusRef: autoFocusRef,
          restoreFocusRef: restoreFocusRef
        }, {
          children: children
        }))
      }))]
    })
  }) : null;
};

export { Modal };
/*  */
