// @oracle/oraclejet-preact/es/UNSAFE_Label.js
/* @oracle/oraclejet-preact: 14.0.4 */
import { _ as __rest } from './tslib.es6-5eaa88c2.js';
import { jsx } from 'preact/jsx-runtime';
import { useFormFieldContext } from './hooks/UNSAFE_useFormFieldContext.js';
import { classNames } from './utils/UNSAFE_classNames.js';
import "./UNSAFE_Label.css";
import { useFormContext } from './hooks/UNSAFE_useFormContext.js';
import 'preact';
import 'preact/hooks';

const labelStyles = {
  base: "cx2le4",
  startBase: "_8g0i7h",
  topBase: "_1qea10i",
  insideBase: "bkjbn7",
  insideReadonlyInEnabledForm: "_1y562h",
  insideReadonlyNotInEnabledForm: "x3m8dg",
  insideNonReadonly: "s6omgl",
  insideNonReadonlyHasValue: "_194slpv",
  insideDisabled: "_98bf7y",
  insideFocused: "_1ee91c",
  insideError: "nxgyma",
  insideWarning: "_1einnx3",
  noWrap: "_1t39xdx"
};

const StyledLabel = _a => {
  var {
    hasValue,
    readonly,
    disabled,
    isFocused,
    variant = 'inside'
  } = _a,
      props = __rest(_a, ["hasValue", "readonly", "disabled", "isFocused", "variant"]);

  const {
    isFormLayout,
    isReadonly: isReadonlyForm,
    labelWrapping
  } = useFormContext();
  const cssClassNames = variant === 'inside' || variant === 'insideError' || variant === 'insideWarning' ? classNames([labelStyles.base, labelStyles.insideBase, readonly && (isFormLayout && !isReadonlyForm ? labelStyles.insideReadonlyInEnabledForm : labelStyles.insideReadonlyNotInEnabledForm), !readonly && labelStyles.insideNonReadonly, !readonly && (hasValue || isFocused) && labelStyles.insideNonReadonlyHasValue, !readonly && isFocused && labelStyles.insideFocused, disabled && labelStyles.insideDisabled, !disabled && !readonly && (variant === 'insideError' || variant === 'insideWarning') && labelStyles[variant]]) : variant === 'start' ? classNames([labelStyles.base, labelStyles.startBase, labelWrapping === 'truncate' && labelStyles.noWrap]) : variant === 'top' ? classNames([labelStyles.base, labelStyles.topBase, labelWrapping === 'truncate' && labelStyles.noWrap]) : undefined; // TODO: We need to decide whether to standardize on 'class' or 'className'.  Otherwise, we'll
  // run into issues destructuring style interpolations, for example having to concatenate both.

  return jsx("label", Object.assign({}, props, {
    class: cssClassNames
  }));
};

const Label = _a => {
  var {
    forId
  } = _a,
      props = __rest(_a, ["forId"]);

  const {
    hasValue,
    isDisabled,
    isFocused,
    isReadonly
  } = useFormFieldContext();
  return jsx(StyledLabel, Object.assign({
    for: forId,
    hasValue: hasValue,
    disabled: isDisabled,
    isFocused: isFocused,
    readonly: isReadonly
  }, props));
};

/**
 * @license
 * Copyright (c) %FIRST_YEAR% %CURRENT_YEAR%, Oracle and/or its affiliates.
 * The Universal Permissive License (UPL), Version 1.0
 * as shown at https://oss.oracle.com/licenses/upl/
 * @ignore
 */

export { Label };
/*  */
