// @oracle/oraclejet-preact/es/UNSAFE_Chip.js
/* @oracle/oraclejet-preact: 14.0.4 */
import { jsx } from 'preact/jsx-runtime';
import { useActionable } from './hooks/UNSAFE_useActionable.js';
import "./UNSAFE_Chip.css";
import { useCallback } from 'preact/hooks';
import { classNames } from './utils/UNSAFE_classNames.js';
import { useTabbableMode } from './hooks/UNSAFE_useTabbableMode.js';
import './utils/UNSAFE_mergeProps.js';
import './hooks/UNSAFE_useHover.js';
import './hooks/UNSAFE_useToggle.js';
import './hooks/UNSAFE_useActive.js';
import './hooks/UNSAFE_useFocus.js';
import './hooks/UNSAFE_usePress.js';
import 'preact';

const cursorandFocusStyle = "_15ygee0";
const styles = {
  base: "_1mph7sn",
  disabled: "_1maz683",
  info: "pd9h3i",
  infoSubtle: "pprnbc",
  infoSubtleLighter: "e9w2ve",
  childrenMargins: "_12atjh4"
};
function Chip({
  isSelected,
  isDisabled,
  accessibleLabel,
  children,
  onToggle
}) {
  const handleAction = useCallback(() => {
    onToggle === null || onToggle === void 0 ? void 0 : onToggle({
      previousValue: isSelected,
      value: !isSelected
    });
  }, [isSelected, onToggle]);
  const {
    actionableProps,
    isHover,
    isActive
  } = useActionable(handleAction);
  const variant = isDisabled ? 'disabled' : isSelected || isActive ? 'info' : isHover ? 'infoSubtle' : 'infoSubtleLighter';
  const classes = classNames([styles.base, isDisabled ? styles.disabled : `${styles.base} ${styles[variant]} ${cursorandFocusStyle}`, styles.childrenMargins]);
  const {
    isTabbable
  } = useTabbableMode();
  return isDisabled ? jsx("div", Object.assign({
    class: classes
  }, {
    children: children
  })) : jsx("div", Object.assign({}, actionableProps, {
    class: classes,
    tabIndex: isTabbable ? 0 : -1,
    "aria-label": accessibleLabel,
    role: "switch",
    "aria-checked": isSelected ? 'true' : 'false'
  }, {
    children: children
  }));
}

export { Chip };
/*  */
