// @oracle/oraclejet-preact/es/UNSAFE_Text.js
/* @oracle/oraclejet-preact: 14.0.4 */
import { _ as __rest } from './tslib.es6-5eaa88c2.js';
import { jsx } from 'preact/jsx-runtime';
import "./UNSAFE_Text.css";
import { mergeInterpolations } from './utils/UNSAFE_mergeInterpolations.js';
import './utils/UNSAFE_classNames.js';
import './_curry1-f2673bab.js';
import './_curry3-6b8e8079.js';
import './_curry2-d7587d88.js';
import './_has-cbbf4e16.js';

const styles = {
  base: "_1j3gtul",
  lineClamp: "_16fdk1x",
  truncation: {
    none: '',
    clip: "_13clz8c",
    ellipsis: "jtef62"
  },
  variant: {
    primary: "_20xmiq",
    secondary: "ay7b1s",
    disabled: "tpaze5",
    danger: "i7g6yo",
    warning: "yoqbwk",
    success: "_1lgwpab",
    inherit: "yatgrc"
  },
  size: {
    '2xs': "_1udekkt",
    xs: "_1qupb0o",
    sm: "_1wviy3z",
    md: "vkd19s",
    lg: "_12d3byw",
    xl: "_1ulujr9",
    inherit: "_45u4c8"
  },
  weight: {
    normal: "_59wm4f",
    semiBold: "_1bxnth8",
    bold: "s8x7zi",
    inherit: "_1mptypd"
  },
  hyphens: {
    auto: "_1ahbq4r",
    none: "_1fm9smf"
  }
};
const interpolations = [() => ({
  class: styles.base
}), ({
  size = 'inherit'
}) => ({
  class: styles.size[size]
}), ({
  weight = 'inherit'
}) => ({
  class: styles.weight[weight]
}), ({
  variant = 'inherit'
}) => ({
  class: styles.variant[variant]
}), ({
  hyphens = 'auto'
}) => ({
  class: styles.hyphens[hyphens]
}), ({
  truncation = 'none'
}) => ({
  class: styles.truncation[truncation]
}), ({
  lineClamp
}) => !lineClamp ? {} : {
  class: styles.lineClamp,
  webkitLineClamp: lineClamp
}];
const styleInterpolations = mergeInterpolations(interpolations);
const childrenStyles = "bs5gdg";
function Text(_a) {
  var {
    children
  } = _a,
      props = __rest(_a, ["children"]);

  const _b = styleInterpolations(props),
        {
    class: cls
  } = _b,
        styles = __rest(_b, ["class"]);

  return jsx("span", Object.assign({
    class: cls,
    style: styles,
    id: props.id
  }, {
    children: jsx("span", Object.assign({
      className: childrenStyles
    }, {
      children: children
    }))
  }));
}

export { Text };
/*  */
