// @oracle/oraclejet-preact/cjs/Heading-97004a9b.js
/* @oracle/oraclejet-preact: 14.0.4 */
'use strict';

var tslib_es6 = require('./tslib.es6-e79a0c43.js');
var jsxRuntime = require('preact/jsx-runtime');
var keys = require('./keys-fcc45418.js');
var utils_UNSAFE_arrayUtils = require('./utils/UNSAFE_arrayUtils.js');

var utils_UNSAFE_mergeInterpolations = require('./utils/UNSAFE_mergeInterpolations.js');

const headingElementTypes = utils_UNSAFE_arrayUtils.stringLiteralArray(['h1', 'h2', 'h3', 'h4', 'h5', 'h6']);
const headingsizeMap = {
  h1: '2xl',
  h2: 'xl',
  h3: 'lg',
  h4: 'md',
  h5: 'sm',
  h6: 'xs'
};
const styles = {
  base: "c8jr8v",
  size: {
    xs: "docxw7",
    sm: "_1jk617g",
    md: "_1npq3ly",
    lg: "na8ijt",
    xl: "p35uav",
    '2xl': "m8728u",
    inherit: "e6hy59"
  },
  childrenMargins: "_1ey2ixr"
};
const sizeOptions = keys.keys_1(styles.size);
const interpolations = [() => ({
  class: styles.base
}), ({
  size = '2xl'
}) => ({
  class: styles.size[size]
})];
const styleInterpolations = utils_UNSAFE_mergeInterpolations.mergeInterpolations(interpolations);
function Heading(_a) {
  var _b;

  var {
    children
  } = _a,
      props = tslib_es6.__rest(_a, ["children"]);

  const Component = props.as; //@ts-ignore

  props.size = (_b = props.size) !== null && _b !== void 0 ? _b : headingsizeMap[Component];

  const _c = styleInterpolations(props),
        {
    class: cls
  } = _c,
        style = tslib_es6.__rest(_c, ["class"]);

  return jsxRuntime.jsx(Component, Object.assign({
    className: cls,
    style: style,
    id: props.id
  }, {
    children: jsxRuntime.jsx("span", Object.assign({
      className: styles.childrenMargins
    }, {
      children: children
    }))
  }));
}

exports.Heading = Heading;
exports.headingElementTypes = headingElementTypes;
exports.sizeOptions = sizeOptions;
/*  */
