// @oracle/oraclejet-preact/cjs/Link-e254783b.js
/* @oracle/oraclejet-preact: 14.0.4 */
'use strict';

var jsxRuntime = require('preact/jsx-runtime');

require('./utils/UNSAFE_classNames.js');
var hooks_UNSAFE_usePress = require('./hooks/UNSAFE_usePress.js');
var classNames = require('./classNames-fe0c9b44.js');

const styles = {
  default: "_5us9ht",
  primary: "_1uf0yzo",
  secondary: "_1v0gjyu",
  standalone: "_10c0p3z",
  embedded: "_16h21b2"
};

const noop = () => {};

function Link({
  href,
  variant = 'default',
  placement = 'standalone',
  target,
  ariaLabelledBy,
  children,
  onClick
}) {
  const linkClass = classNames.classNames([styles.default, styles[placement], variant !== 'default' && styles[variant]]);
  const {
    pressProps
  } = hooks_UNSAFE_usePress.usePress(onClick !== null && onClick !== void 0 ? onClick : noop);
  return jsxRuntime.jsx("a", Object.assign({
    href: href,
    "aria-labelledby": ariaLabelledBy,
    target: target
  }, pressProps, {
    class: linkClass
  }, {
    children: children
  }));
}

exports.Link = Link;
/*  */
