// @oracle/oraclejet-preact/cjs/MessageSummary-b3abe0a3.js
/* @oracle/oraclejet-preact: 14.0.4 */
'use strict';

var jsxRuntime = require('preact/jsx-runtime');

require('./utils/UNSAFE_classNames.js');
var classNames = require('./classNames-fe0c9b44.js');

const messageSummaryStyles = {
  // TODO: Reevaluate once the TEXT component is available (JET-46891)
  base: "_3olrzv",
  banner: "_9fzsr3",
  inline: "j541z9",
  toast: "z6vqbn"
};
/**
 * Summary Component for rendering the summary text of the Message
 */

function MessageSummary({
  text,
  variant = 'banner'
}) {
  // If detail content is rendered, then wrap it in a div with specified style classes
  // TODO: JET-48715 - remove marker classes
  // Right now, oj-message-banner component uses this for unit tests and has no alternate way for
  // testing. This will need to be removed as we work on oj-c-message-banner.
  const markerClassForTests = 'oj-c-messagebanner-summary';
  const classes = classNames.classNames([variant === 'banner' && markerClassForTests, messageSummaryStyles.base, messageSummaryStyles[variant]]);
  return jsxRuntime.jsx("div", Object.assign({
    role: "heading",
    class: classes
  }, {
    children: text
  }));
}

exports.MessageSummary = MessageSummary;
/*  */
