// @oracle/oraclejet-preact/cjs/MessageLayer-12cbabd5.js
/* @oracle/oraclejet-preact: 14.0.4 */
'use strict';

var jsxRuntime = require('preact/jsx-runtime');

var hooks_UNSAFE_useUser = require('./hooks/UNSAFE_useUser.js');
var utils_UNSAFE_arrayUtils = require('./utils/UNSAFE_arrayUtils.js');
require('./utils/UNSAFE_classNames.js');
require('./UNSAFE_Flex.js');
var UNSAFE_WindowOverlay = require('./UNSAFE_WindowOverlay.js');
var classNames = require('./classNames-fe0c9b44.js');
var Flex = require('./Flex-70f45c1b.js');

const positions = utils_UNSAFE_arrayUtils.stringLiteralArray(['top', 'top-left', 'top-right', 'top-start', 'top-end', 'bottom', 'bottom-left', 'bottom-right', 'bottom-start', 'bottom-end']);
const styles = {
  notificationPosition: "_1mfywjs",
  ['top-left']: "s4p4w3",
  top: "_1dl9lw1",
  ['top-right']: "_1x0wkpb",
  ['bottom-left']: "yj0man",
  bottom: "_1j4iogf",
  ['bottom-right']: "_1pq7mcp"
};

const MessageLayer = ({
  children,
  offset,
  testId,
  position = 'bottom'
}) => {
  var _a, _b;

  const {
    direction
  } = hooks_UNSAFE_useUser.useUser();
  const isLtr = direction === 'ltr';
  let logicalPosition;

  switch (position) {
    case 'top-start':
      logicalPosition = isLtr ? styles['top-left'] : styles['top-right'];
      break;

    case 'top-end':
      logicalPosition = isLtr ? styles['top-right'] : styles['top-left'];
      break;

    case 'bottom-start':
      logicalPosition = isLtr ? styles['bottom-left'] : styles['bottom-right'];
      break;

    case 'bottom-end':
      logicalPosition = isLtr ? styles['bottom-right'] : styles['bottom-left'];
      break;

    default:
      logicalPosition = styles[position];
      break;
  }

  const positionClasses = classNames.classNames([styles.notificationPosition, logicalPosition]);
  const offsetStyle = typeof offset === 'number' ? {
    '--oj-c-PRIVATE-DO-NOT-USE-notification-vertical-offset': `${offset}px`,
    '--oj-c-PRIVATE-DO-NOT-USE-notification-horizontal-offset': `${offset}px`
  } : {
    '--oj-c-PRIVATE-DO-NOT-USE-notification-vertical-offset': `${(_a = offset === null || offset === void 0 ? void 0 : offset.vertical) !== null && _a !== void 0 ? _a : 0}px`,
    '--oj-c-PRIVATE-DO-NOT-USE-notification-horizontal-offset': `${(_b = offset === null || offset === void 0 ? void 0 : offset.horizontal) !== null && _b !== void 0 ? _b : 0}px`
  };
  return jsxRuntime.jsx(UNSAFE_WindowOverlay.WindowOverlay, {
    children: jsxRuntime.jsx("div", Object.assign({
      class: positionClasses,
      "data-testid": testId,
      style: offset !== undefined ? offsetStyle : undefined
    }, {
      children: jsxRuntime.jsx(Flex.Flex, Object.assign({
        direction: "column",
        gap: "2x"
      }, {
        children: children
      }))
    }))
  });
};

exports.MessageLayer = MessageLayer;
exports.positions = positions;
/*  */
