// @oracle/oraclejet-preact/cjs/MessageTimestamp-bcf4ac8d.js
/* @oracle/oraclejet-preact: 14.0.4 */
'use strict';

var jsxRuntime = require('preact/jsx-runtime');

require('./utils/UNSAFE_classNames.js');
var MessageFormattingUtils = require('./MessageFormattingUtils-55602416.js');
var classNames = require('./classNames-fe0c9b44.js');

const messageTimestampStyles = {
  banner: "ojtty7",
  inline: undefined,
  toast: undefined
};
/**
 * Timestamp Component for rendering timestamp in Message
 */

function MessageTimestamp({
  value,
  variant = 'banner'
}) {
  // If detail content is rendered, then wrap it in a div with specified style classes
  // TODO: JET-48715 - remove marker classes
  // Right now, oj-message-banner component uses this for unit tests and has no alternate way for
  // testing. This will need to be removed as we work on oj-c-message-banner.
  const markerClassForTests = 'oj-c-messagebanner-timestamp';
  const classes = classNames.classNames([variant === 'banner' && markerClassForTests, messageTimestampStyles[variant]]); // Otherwise, render the timestamp

  const formattedTimestamp = MessageFormattingUtils.formatTimestamp(value);
  return jsxRuntime.jsx("div", Object.assign({
    class: classes
  }, {
    children: formattedTimestamp
  }));
}

exports.MessageTimestamp = MessageTimestamp;
/*  */
