// @oracle/oraclejet-preact/cjs/MessageCloseButton-a750b114.js
/* @oracle/oraclejet-preact: 14.0.4 */
'use strict';

var jsxRuntime = require('preact/jsx-runtime');

require('./UNSAFE_Button.js');
require('./index-7cf551b0.js');
require('./utils/UNSAFE_classNames.js');
var Button = require('./Button-fe40b39e.js');
var UNSAFE_Icons = require('./index-fbc2a690.js');
var classNames = require('./classNames-fe0c9b44.js');

const messageCloseButtonStyles = {
  base: "_1jgzsnu",
  banner: "_15t0p4o",
  inline: undefined,
  toast: "_1ex2abo"
};
/**
 * A Component for rendering the message close button
 */

function MessageCloseButton({
  onAction,
  buttonRenderer,
  title = 'Close',
  variant = 'banner'
}) {
  const renderedButton = buttonRenderer ? buttonRenderer(title, onAction, variant) : jsxRuntime.jsx(Button.Button, {
    startIcon: jsxRuntime.jsx(UNSAFE_Icons.SvgIcoClose, {}),
    size: "sm",
    title: title,
    variant: "borderless",
    onAction: onAction
  }); // TODO: JET-48715 - remove marker classes
  // Right now, oj-message-banner component uses this for unit tests and has no alternate way for
  // testing. This will need to be removed as we work on oj-c-message-banner.

  const markerClassForTests = 'oj-c-messagebanner-close-icon';
  const classes = classNames.classNames([variant === 'banner' && markerClassForTests, messageCloseButtonStyles.base, messageCloseButtonStyles[variant]]); // Otherwise, render the close icon

  return jsxRuntime.jsx("div", Object.assign({
    class: classes
  }, {
    children: renderedButton
  }));
}

exports.MessageCloseButton = MessageCloseButton;
/*  */
