// @oracle/oraclejet-preact/cjs/MessageDetail-7665d979.js
/* @oracle/oraclejet-preact: 14.0.4 */
'use strict';

var jsxRuntime = require('preact/jsx-runtime');

var preact = require('preact');
require('./utils/UNSAFE_classNames.js');
var MessageFormattingUtils = require('./MessageFormattingUtils-55602416.js');
var classNames = require('./classNames-fe0c9b44.js');

const messageDetailStyles = {
  // TODO: Reevaluate once the TEXT component is available (JET-46891)
  base: "n6tggq",
  banner: "_16w2c1a",
  inline: undefined,
  toast: "_1pi1655"
};
/**
 * Default renderer for rendering the detail content.
 *
 * @param item The data item object
 * @returns Rendered detail content
 */

function defaultDetailRenderer(item) {
  const {
    detail
  } = item.data; // If the detail is null or an empty string, do not render the
  // content row

  if (!MessageFormattingUtils.isValidValueForProp(detail)) {
    return null;
  }

  return jsxRuntime.jsx(preact.Fragment, {
    children: detail
  });
}
/**
 * Detail Component for rendering the detail content of the Message
 */


function MessageDetail({
  item,
  renderer = defaultDetailRenderer,
  variant = 'banner'
}) {
  const renderedContent = renderer(item);
  if (renderedContent == null) return null; // If detail content is rendered, then wrap it in a div with specified style classes
  // TODO: JET-48715 - remove marker classes
  // Right now, oj-message-banner component uses this for unit tests and has no alternate way for
  // testing. This will need to be removed as we work on oj-c-message-banner.

  const markerClassForTests = 'oj-c-messagebanner-detail';
  const classes = classNames.classNames([variant === 'banner' && markerClassForTests, messageDetailStyles.base, messageDetailStyles[variant]]);
  return jsxRuntime.jsx("div", Object.assign({
    class: classes
  }, {
    children: renderedContent
  }));
}

exports.MessageDetail = MessageDetail;
/*  */
