// @oracle/oraclejet-preact/cjs/Modal-3c05f90a.js
/* @oracle/oraclejet-preact: 14.0.4 */
'use strict';

var jsxRuntime = require('preact/jsx-runtime');
var compat = require('preact/compat');
require('./UNSAFE_FocusTrap.js');
var UNSAFE_Layer = require('./UNSAFE_Layer.js');

var FocusTrap = require('./FocusTrap-0c3e0eaf.js');

/**
 *
 * @param props
 */
const useModal = (props) => {
    const { isOpen, onBackdropClick: onBackdropClickProp, } = props;
    const modalRef = compat.useRef(null);
    const mouseDownTarget = compat.useRef(null);
    const onMouseDown = compat.useCallback((event) => {
        if (modalRef.current === event.target) {
            mouseDownTarget.current = event.target;
        }
    }, []);
    const onBackdropClick = compat.useCallback((event) => {
        if (modalRef.current === event.target) {
            event.stopPropagation();
            // event should originate on the same target
            if (mouseDownTarget.current === event.target) {
                onBackdropClickProp === null || onBackdropClickProp === void 0 ? void 0 : onBackdropClickProp();
            }
        }
    }, [onBackdropClickProp]);
    compat.useEffect(() => {
        const modal = modalRef.current;
        if (modal) {
            modal.addEventListener('mousedown', onMouseDown);
            modal.addEventListener('click', onBackdropClick);
            return () => {
                modal.removeEventListener('mousedown', onMouseDown);
                modal.removeEventListener('click', onBackdropClick);
            };
        }
        return () => { };
    }, [isOpen, onMouseDown, onBackdropClick]);
    return {
        modalRef,
    };
};

const styles$2 = {
  base: "_12yz8ja"
};

const ModalBackdrop = () => jsxRuntime.jsx("div", {
  class: styles$2.base
});

const styles$1 = {
  base: "_1vxwng5"
};
const ModalContainer = compat.forwardRef(({
  children
}, ref) => jsxRuntime.jsx("div", Object.assign({
  class: styles$1.base,
  ref: ref
}, {
  children: children
})));
ModalContainer.displayName = 'Forwarded<ModalContainer>';

const styles = {
  base: "_1wzau58"
};

const ModalWrapper = ({
  children
}) => jsxRuntime.jsx("div", Object.assign({
  class: styles.base
}, {
  children: children
}));

const Modal = ({
  autoFocusRef,
  children,
  isOpen,
  onBackdropClick,
  restoreFocusRef
}) => {
  const {
    modalRef
  } = useModal({
    isOpen,
    onBackdropClick
  });
  return isOpen ? jsxRuntime.jsx(UNSAFE_Layer.Layer, {
    children: jsxRuntime.jsxs(ModalWrapper, {
      children: [jsxRuntime.jsx(ModalBackdrop, {}), jsxRuntime.jsx(ModalContainer, Object.assign({
        ref: modalRef
      }, {
        children: jsxRuntime.jsx(FocusTrap.FocusTrap, Object.assign({
          autoFocusRef: autoFocusRef,
          restoreFocusRef: restoreFocusRef
        }, {
          children: children
        }))
      }))]
    })
  }) : null;
};

exports.Modal = Modal;
/*  */
