// @oracle/oraclejet-preact/cjs/SelectMenuItem-ac10e379.js
/* @oracle/oraclejet-preact: 14.0.4 */
'use strict';

var jsxRuntime = require('preact/jsx-runtime');
var hooks = require('preact/hooks');
require('./index-7cf551b0.js');
var MenuItem = require('./MenuItem-2928619f.js');
var preact = require('preact');
var UNSAFE_Icons = require('./index-fbc2a690.js');

/**
 * @license
 * Copyright (c) %FIRST_YEAR% %CURRENT_YEAR%, Oracle and/or its affiliates.
 * Licensed under The Universal Permissive License (UPL), Version 1.0
 * as shown at https://oss.oracle.com/licenses/upl/
 * @ignore
 */
const SelectMenuGroupContext = preact.createContext({
    isMultiple: false
});

/**
 * @license
 * Copyright (c) %FIRST_YEAR% %CURRENT_YEAR%, Oracle and/or its affiliates.
 * Licensed under The Universal Permissive License (UPL), Version 1.0
 * as shown at https://oss.oracle.com/licenses/upl/
 * @ignore
 */
/**
 * Utility hook for consuming the MenuContext
 *
 * @returns The value of MenuContext provider
 */
function useSelectMenuGroupContext() {
    return hooks.useContext(SelectMenuGroupContext);
}

function SelectMenuItem({ value, label, isDisabled, endIcon }) {
    const { isMultiple, value: selectedValue, onCommit } = useSelectMenuGroupContext();
    const { isSelected, memoizedOnCommit } = hooks.useMemo(() => {
        if (!isMultiple) {
            return {
                isSelected: selectedValue === value,
                memoizedOnCommit: () => {
                    onCommit === null || onCommit === void 0 ? void 0 : onCommit({ value, previousValue: selectedValue });
                }
            };
        }
        return {
            isSelected: !!(selectedValue === null || selectedValue === void 0 ? void 0 : selectedValue.includes(value)),
            memoizedOnCommit: () => {
                const proposedValue = isSelected
                    ? (selectedValue === null || selectedValue === void 0 ? void 0 : selectedValue.filter((val) => val !== value)) || []
                    : [...(selectedValue || []), value];
                onCommit === null || onCommit === void 0 ? void 0 : onCommit({ value: proposedValue, previousValue: selectedValue });
            }
        };
    }, [selectedValue, value, onCommit, isMultiple]);
    return (jsxRuntime.jsx(MenuItem.BaseMenuItem, { role: isMultiple ? 'menuitemcheckbox' : 'menuitemradio', isChecked: isSelected, isDisabled: isDisabled, startIcon: isSelected ? jsxRuntime.jsx(UNSAFE_Icons.SvgIcoCheck, {}) : jsxRuntime.jsx("span", {}), label: label, onAction: memoizedOnCommit, endIcon: endIcon }));
}

exports.SelectMenuGroupContext = SelectMenuGroupContext;
exports.SelectMenuItem = SelectMenuItem;
/*  */
