// @oracle/oraclejet-preact/cjs/SplitMenuButton-8c71339c.js
/* @oracle/oraclejet-preact: 14.0.4 */
'use strict';

var tslib_es6 = require('./tslib.es6-e79a0c43.js');
var jsxRuntime = require('preact/jsx-runtime');
require('./UNSAFE_Button.js');
require('./index-7cf551b0.js');
require('./UNSAFE_Menu.js');
require('./UNSAFE_BaseButton.js');
require('./utils/UNSAFE_classNames.js');
var hooks_UNSAFE_useTabbableMode = require('./hooks/UNSAFE_useTabbableMode.js');
var utils_UNSAFE_interpolations_dimensions = require('./utils/UNSAFE_interpolations/dimensions.js');
var utils_UNSAFE_mergeInterpolations = require('./utils/UNSAFE_mergeInterpolations.js');

var compat = require('preact/compat');
var hooks = require('preact/hooks');
var classNames = require('./classNames-fe0c9b44.js');
var BaseButton = require('./BaseButton-1c38289b.js');
var Button = require('./Button-fe40b39e.js');
var UNSAFE_Icons = require('./index-fbc2a690.js');
var Menu = require('./Menu-7bc3a21d.js');

const compStyles = {
  divider: "_13yktiu",
  disabledDivider: "_1n1ww93",
  icon: "xwx0yp",
  root: "_1i85y6j",
  childrenMargins: "_1y831ki"
};
const menuButtonStyling = ['embedded', 'min', 'noBorderRadiusStart'];
const menuButtonActiveStyling = [...menuButtonStyling, 'active'];
const actionbuttonStyling = ['embedded', 'fill', 'noBorderRadiusEnd'];
const actionButtonActiveStyling = [...actionbuttonStyling, 'active'];
const interpolations = [...Object.values(utils_UNSAFE_interpolations_dimensions.dimensionInterpolations)];
const styleInterpolations = utils_UNSAFE_mergeInterpolations.mergeInterpolations(interpolations);
/**
 * A SplitMenuButton is a combined action and menu button.
 */

const SplitMenuButton = compat.forwardRef((_a, ref) => {
  var {
    children,
    label = '',
    variant = 'outlined',
    isDisabled = false,
    size = 'md',
    onAction
  } = _a,
      props = tslib_es6.__rest(_a, ["children", "label", "variant", "isDisabled", "size", "onAction"]);

  const [isMenuOpen, setIsMenuOpen] = hooks.useState(false);
  const [isActive, setIsActive] = hooks.useState(false);

  const _b = styleInterpolations(props),
        {
    class: cls
  } = _b,
        styles = tslib_es6.__rest(_b, ["class"]); // SPACE or ENTER on root area invokes action, as does CLICK/TAP on LEFT SIDE ACTION AREA
  // DOWN ARROW invokes menu, as does CLICK/TAP on RIGHT SIDE MENU AREA


  const onKeyDown = hooks.useCallback(event => {
    if (event.key === ' ' || event.key === 'Enter') {
      setIsActive(true);
    } else if (event.key == 'ArrowDown') {
      !isDisabled && setIsMenuOpen(menuOpen => !menuOpen); // Prevent page scroll for down arrow, user definitely wanted to show menu in this case.
      // Don't prevent other keypress defaults, as user may want to hold SPACE down to scroll page.

      event.preventDefault();
    }
  }, [isDisabled, setIsMenuOpen, setIsActive]);
  const onKeyUp = hooks.useCallback(event => {
    if (event.key === ' ' || event.key === 'Enter') {
      !isDisabled && onAction && onAction();
      setIsActive(false);
    }
  }, [isDisabled, onAction, setIsActive]);
  const onBlur = hooks.useCallback(() => {
    setIsActive(false);
  }, [setIsActive]);

  const _ref = hooks.useRef(null);

  hooks.useImperativeHandle(ref, () => _ref.current, [_ref]);

  const menuHandler = () => {
    setIsMenuOpen(!isMenuOpen);
  };

  const handleClose = e => {
    var _a;

    if (e.reason === 'cancel' || 'focusLoss') {
      (_a = _ref.current) === null || _a === void 0 ? void 0 : _a.focus();
    }

    setIsMenuOpen(false);
  };

  const dividerStyles = classNames.classNames([compStyles.divider, isDisabled && compStyles.disabledDivider]);
  const rootStyles = classNames.classNames([compStyles.root]); // Using Span for top button allows us to avoid :hover state rendering by
  // taking advantage of :enabled not on spans

  const rootButton = () => {
    return jsxRuntime.jsx(BaseButton.BaseButton, Object.assign({
      elementDetails: {
        type: 'span'
      },
      ref: _ref,
      isDisabled: isDisabled,
      styling: ['container', 'min'],
      variant: variant,
      size: size,
      accessibleLabel: label,
      ariaRoledescription: "split menu button, press down arrow to invoke menu",
      ariaExpanded: isMenuOpen
    }, {
      children: jsxRuntime.jsxs(hooks_UNSAFE_useTabbableMode.TabbableModeContext.Provider, Object.assign({
        value: {
          isTabbable: false
        }
      }, {
        children: [jsxRuntime.jsx(BaseButton.BaseButton, Object.assign({
          elementDetails: {
            type: 'span'
          },
          variant: variant,
          styling: getActionButtonStyling(isActive),
          size: size,
          isDisabled: isDisabled,
          ariaHidden: true,
          onAction: onAction
        }, {
          children: label
        })), jsxRuntime.jsx("div", {
          class: `${dividerStyles}`
        }), jsxRuntime.jsx(BaseButton.BaseButton, Object.assign({
          elementDetails: {
            type: 'span'
          },
          variant: variant,
          styling: getMenuButtonStyling(isMenuOpen),
          size: size,
          isDisabled: isDisabled,
          ariaHidden: true,
          onAction: menuHandler
        }, {
          children: jsxRuntime.jsx(Button.ButtonLayout, {
            display: "icons",
            startIcon: jsxRuntime.jsx(UNSAFE_Icons.SvgIcoNavDown, {}),
            size: size,
            styling: "embedded"
          })
        }))]
      }))
    }));
  };

  if (isDisabled) {
    return jsxRuntime.jsx("span", Object.assign({
      role: "toolbar",
      class: `${rootStyles} ${cls}`,
      style: styles
    }, {
      children: rootButton()
    }));
  } else {
    return jsxRuntime.jsxs("span", Object.assign({
      role: "toolbar",
      class: `${rootStyles} ${cls}`,
      style: styles,
      onKeyDown: !isDisabled && onKeyDown,
      onKeyUp: onKeyUp,
      onBlur: onBlur
    }, {
      children: [rootButton(), jsxRuntime.jsx(Menu.Menu, Object.assign({
        anchorRef: _ref,
        isOpen: isMenuOpen,
        onClose: handleClose
      }, {
        children: jsxRuntime.jsx("span", Object.assign({
          className: compStyles.childrenMargins
        }, {
          children: children
        }))
      }))]
    }));
  }
});

function getMenuButtonStyling(isActive) {
  return isActive ? menuButtonActiveStyling : menuButtonStyling;
}

function getActionButtonStyling(isActive) {
  return isActive ? actionButtonActiveStyling : actionbuttonStyling;
}

exports.SplitMenuButton = SplitMenuButton;
/*  */
