// @oracle/oraclejet-preact/cjs/Text-4b71357e.js
/* @oracle/oraclejet-preact: 14.0.4 */
'use strict';

var tslib_es6 = require('./tslib.es6-e79a0c43.js');
var jsxRuntime = require('preact/jsx-runtime');

var utils_UNSAFE_mergeInterpolations = require('./utils/UNSAFE_mergeInterpolations.js');

const styles = {
  base: "_1j3gtul",
  lineClamp: "_16fdk1x",
  truncation: {
    none: '',
    clip: "_13clz8c",
    ellipsis: "jtef62"
  },
  variant: {
    primary: "_20xmiq",
    secondary: "ay7b1s",
    disabled: "tpaze5",
    danger: "i7g6yo",
    warning: "yoqbwk",
    success: "_1lgwpab",
    inherit: "yatgrc"
  },
  size: {
    '2xs': "_1udekkt",
    xs: "_1qupb0o",
    sm: "_1wviy3z",
    md: "vkd19s",
    lg: "_12d3byw",
    xl: "_1ulujr9",
    inherit: "_45u4c8"
  },
  weight: {
    normal: "_59wm4f",
    semiBold: "_1bxnth8",
    bold: "s8x7zi",
    inherit: "_1mptypd"
  },
  hyphens: {
    auto: "_1ahbq4r",
    none: "_1fm9smf"
  }
};
const interpolations = [() => ({
  class: styles.base
}), ({
  size = 'inherit'
}) => ({
  class: styles.size[size]
}), ({
  weight = 'inherit'
}) => ({
  class: styles.weight[weight]
}), ({
  variant = 'inherit'
}) => ({
  class: styles.variant[variant]
}), ({
  hyphens = 'auto'
}) => ({
  class: styles.hyphens[hyphens]
}), ({
  truncation = 'none'
}) => ({
  class: styles.truncation[truncation]
}), ({
  lineClamp
}) => !lineClamp ? {} : {
  class: styles.lineClamp,
  webkitLineClamp: lineClamp
}];
const styleInterpolations = utils_UNSAFE_mergeInterpolations.mergeInterpolations(interpolations);
const childrenStyles = "bs5gdg";
function Text(_a) {
  var {
    children
  } = _a,
      props = tslib_es6.__rest(_a, ["children"]);

  const _b = styleInterpolations(props),
        {
    class: cls
  } = _b,
        styles = tslib_es6.__rest(_b, ["class"]);

  return jsxRuntime.jsx("span", Object.assign({
    class: cls,
    style: styles,
    id: props.id
  }, {
    children: jsxRuntime.jsx("span", Object.assign({
      className: childrenStyles
    }, {
      children: children
    }))
  }));
}

exports.Text = Text;
/*  */
