// @oracle/oraclejet-preact/cjs/UNSAFE_Avatar.js
/* @oracle/oraclejet-preact: 14.0.4 */
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var tslib_es6 = require('./tslib.es6-e79a0c43.js');
var jsxRuntime = require('preact/jsx-runtime');
require('./utils/UNSAFE_classNames.js');
require("identity-obj-proxy");
var classNames = require('./classNames-fe0c9b44.js');

/**
 * An avatar represents a person or entity as initials or an image.
 */

function Avatar(_a) {
  var {
    children,
    src
  } = _a,
      otherProps = tslib_es6.__rest(_a, ["children", "src"]);

  if (src && !children) {
    return jsxRuntime.jsx(ImageAvatar, Object.assign({
      src: src
    }, otherProps));
  } else {
    return jsxRuntime.jsx(ContentAvatar, Object.assign({}, otherProps, {
      children: children
    }));
  }
}

const ImageAvatar = _a => {
  var {
    src
  } = _a,
      otherProps = tslib_es6.__rest(_a, ["src"]);

  const classes = classNames.classNames([contentStyles.base, contentStyles.image]);
  return jsxRuntime.jsx(AvatarWrapper, Object.assign({}, otherProps, {
    children: jsxRuntime.jsx("div", {
      class: classes,
      style: {
        backgroundImage: `url(${src})`
      }
    })
  }));
};

const ContentAvatar = _a => {
  var {
    children,
    initials
  } = _a,
      otherProps = tslib_es6.__rest(_a, ["children", "initials"]);

  const classes = [contentStyles.base, contentStyles.content, initials ? contentStyles.initials : contentStyles.icon, !initials && childrenStyles];
  const childContent = initials || children;
  return jsxRuntime.jsx(AvatarWrapper, Object.assign({}, otherProps, {
    children: jsxRuntime.jsx("div", Object.assign({
      "aria-hidden": "true",
      class: classNames.classNames(classes)
    }, {
      children: childContent
    }))
  }));
};

const AvatarWrapper = ({
  background = 'neutral',
  size = 'md',
  shape = 'square',
  children,
  accessibleLabel
}) => {
  const classes = classNames.classNames([styles.base, `oj-c-avatar-bg-${background}`, `oj-c-avatar-${size}`, shape === 'circle' && styles.circle]);

  if (accessibleLabel) {
    return jsxRuntime.jsx("div", Object.assign({
      "aria-label": accessibleLabel,
      role: "img",
      class: classes
    }, {
      children: children
    }));
  } else {
    return jsxRuntime.jsx("div", Object.assign({
      class: classes
    }, {
      children: children
    }));
  }
};

const styles = {
  base: "_1wcqaao",
  circle: "xi8wk0"
};
const childrenStyles = "fso5c2";
const contentStyles = {
  base: "l9udmj",
  content: "_1xi39ou",
  image: "lv68bk",
  initials: "_17vikip",
  icon: "_11iwnhw"
};

exports.Avatar = Avatar;
/*  */
