// @oracle/oraclejet-preact/cjs/UNSAFE_LabelValueLayout.js
/* @oracle/oraclejet-preact: 14.0.4 */
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var jsxRuntime = require('preact/jsx-runtime');
require('./utils/UNSAFE_classNames.js');
require("identity-obj-proxy");
require('./UNSAFE_Flex.js');
var hooks_UNSAFE_useFormContext = require('./hooks/UNSAFE_useFormContext.js');
var hooks_UNSAFE_useFormFieldContext = require('./hooks/UNSAFE_useFormFieldContext.js');
var utils_UNSAFE_size = require('./utils/UNSAFE_size.js');
var classNames = require('./classNames-fe0c9b44.js');
var Flex = require('./Flex-70f45c1b.js');
require('./tslib.es6-e79a0c43.js');
require('./utils/UNSAFE_interpolations/dimensions.js');
require('./utils/UNSAFE_arrayUtils.js');
require('./_curry1-8dba47c0.js');
require('./utils/UNSAFE_mergeInterpolations.js');
require('./_curry3-a11ae74a.js');
require('./_curry2-35fe759b.js');
require('./_has-f646b7fa.js');
require('./utils/UNSAFE_interpolations/boxalignment.js');
require('./keys-fcc45418.js');
require('./utils/UNSAFE_interpolations/flexbox.js');
require('./flexbox-e7360e12.js');
require('./utils/UNSAFE_interpolations/flexitem.js');
require('./flexitem-21bb492c.js');
require('preact');
require('preact/hooks');

const labelSlotStyles = {
  base: "aci6fd",
  start: "_1horddl",
  top: "_1p44q6n",
  topPureReadonly: "_1m57ttq"
};
const labelInnerStyles = {
  // base is labelEdge 'top'
  base: "_5ehzov",
  start: "_1ilc7kb",
  startNotPureReadonly: "_1eg19hs",
  // TODO: Use Text Component instead of CSS
  noWrap: "_2vxp5u"
};
const valueSlotStyles = "_18nipp6";
const LabelValueLayout = ({
  label,
  labelEdge,
  children,
  labelStartWidth = '33%'
}) => {
  const width = utils_UNSAFE_size.sizeToCSS(labelStartWidth);
  const valueWidth = `calc(100% - ${width})`;
  const labelStyles = labelEdge === 'start' ? {
    flexBasis: width,
    width: width,
    maxWidth: width
  } : {};
  const valueStyles = labelEdge === 'start' ? {
    flexBasis: valueWidth,
    width: valueWidth,
    maxWidth: valueWidth
  } : {};
  const {
    isFormLayout,
    isReadonly: isFormReadonly,
    labelWrapping
  } = hooks_UNSAFE_useFormContext.useFormContext();
  const {
    isReadonly
  } = hooks_UNSAFE_useFormFieldContext.useFormFieldContext();
  const isPureReadonly = isFormLayout && isFormReadonly || !isFormLayout && isReadonly;
  const labelStyleClasses = classNames.classNames([labelSlotStyles.base, labelEdge === 'start' && labelSlotStyles.start, // labelEdge === 'start' && isPureReadonly && labelSlotStyles.startReadonly,
  // labelEdge === 'start' && !isPureReadonly && labelSlotStyles.startNotPureReadonly,
  labelEdge === 'top' && labelSlotStyles.top, labelEdge === 'top' && isPureReadonly && labelSlotStyles.topPureReadonly]);
  const labelInnerClasses = classNames.classNames([labelInnerStyles.base, labelEdge === 'start' && labelInnerStyles.start, labelEdge === 'start' && !isPureReadonly && labelInnerStyles.startNotPureReadonly, labelWrapping === 'truncate' && labelInnerStyles.noWrap]);
  return jsxRuntime.jsxs(Flex.Flex, Object.assign({
    wrap: "wrap",
    align: "start"
  }, {
    children: [jsxRuntime.jsx("div", Object.assign({
      class: labelStyleClasses,
      style: labelStyles
    }, {
      children: jsxRuntime.jsx("div", Object.assign({
        class: labelInnerClasses
      }, {
        children: label
      }))
    })), jsxRuntime.jsx("div", Object.assign({
      class: valueSlotStyles,
      style: valueStyles
    }, {
      children: children
    }))]
  }));
};

/**
 * @license
 * Copyright (c) %FIRST_YEAR% %CURRENT_YEAR%, Oracle and/or its affiliates.
 * The Universal Permissive License (UPL), Version 1.0
 * as shown at https://oss.oracle.com/licenses/upl/
 * @ignore
 */

exports.LabelValueLayout = LabelValueLayout;
/*  */
